### PR TITLE
`Integrated code lifecycle`: Show ssh fingerprints

### PR DESCRIPTION
**Pull request by SimonEntholzer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.cit.tum.de/dev/guidelines/performance/) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the principle of **data economy** for all client-server REST calls.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Resolves https://github.com/ls1intum/Artemis/issues/9527 

### Description

This PR adds a subpage to the SSH settings, to let users display Artemis key fingerprints

### Steps for Testing

1. Login and navigate to the SSH settings. 
2. Click the fingerprints link
3. Verify the displayed fingerprint is correct (e.g. delete your known hosts file/remove the test server's entry, and use SSH to clone something)

image

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| ssh-user-settings-fingerprints.component.ts | 100% | ✅ |
| ssh-user-settings-fingerprints.service.ts | 100% | ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| SshFingerprintsProviderService.java | 100% | ✅ |
| SshFingerprintsProviderResource.java | 100% | ✅ |

### Screenshots

image
(Shows fingerprints depending on which are configured for the server)
image

## Release Notes

- **New Features**
  - Introduced a service for managing SSH fingerprints.
  - Added a REST controller for retrieving SSH fingerprints via a new API endpoint.
  - Implemented a user interface component for displaying SSH fingerprints in user settings.
  - Enhanced SSH settings with new routes for managing SSH keys and fingerprints.
  - Expanded user settings with new fields and error messages related to SSH key management.
  - Introduced automatic expiration settings for SSH keys.

- **Bug Fixes**
  - Improved error handling for SSH key management.

- **Documentation**
  - Updated user settings documentation with new error messages and details about SSH keys.

- **Tests**
  - Added comprehensive test suites for the new components and services related to SSH fingerprint management.
  - Established tests for the functionality of retrieving SSH fingerprints.

### CODE REVIEW

**Review by iyannsch — CHANGES_REQUESTED** [review]
Would you like to base this PR on `feature/localVC/support-multiple-SSH-keys`? In here are most "multi SSH keys" LOCs making it hard to review ☺️

**Comment by SimonEntholzer** [comment]
> Would you like to base this PR on `feature/localVC/support-multiple-SSH-keys`? In here are most "multi SSH keys" LOCs making it hard to review ☺️

Yeah I'm waiting to get https://github.com/ls1intum/Artemis/pull/9478 merged, because the finger prints PR depends on the UI changes from that PR, hence it's still draft

**Review by BBesrour — APPROVED** [review]
Tested in a testing session.
Doesn't work properly on test server since they use fall back keys. But Worked fine locally.

**Review by coolchock — DISMISSED** [review]
Tested in a testing session. LGTM 👍

**Review by EneaGore — DISMISSED** [review]
Tested in testing session.

**Review by magkue — COMMENTED** [review]
Tested locally and on TS5. Works fine. Small suggestion. I would call the button "Show fingerprints" or even "Show SSH fingerprints" as this more consistent with the Add button and makes the action of the button clearer.

**Review by magkue — APPROVED** [review]

File: src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/ssh/SshFingerprintsProviderService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+     * @return a map containing the SSH key fingerprints, where the key is the algorithm
+     *         of the public key and the value is its SHA-256 fingerprint.
+     * @throws BadRequestException if there is an error loading keys from the SSH server.
+     */
+    public Map<String, String> getSshFingerPrints() {
+        Map<String, String> fingerprints = new HashMap<>();
+        KeyPairProvider keyPairProvider = sshServer.getKeyPairProvider();
+        if (keyPairProvider != null) {
+            try {
+                keyPairProvider.loadKeys(null).iterator()
```
Comment (iyannsch) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/ssh/SshFingerprintsProviderService.java:46]: Is the `loadKeys` purely to get the iterator? `null` due to the sshServer already containing keys?

File: src/main/webapp/i18n/de/userSettings.json [hunk 2]
```diff
@@ -55,6 +55,10 @@
                 "expiresOn": "Läuft ab am",
                 "hasExpiredOn": "Abgelaufen am",
                 "fingerprint": "Fingerabdruck",
+                "fingerprints": "SSH Fingerabdrücke anzeigen",
+                "sshFingerprints": "SSH Fingerabdrücke",
+                "fingerprintsExplanation": "Mit SSH-Schlüsseln kannst du eine sichere Verbindung zwischen deinem Computer und Artemis herstellen. SSH-Fingerabdrücke stellen sicher, dass der Client eine Verbindung zum richtigen Host herstellt.",
```
Comment (iyannsch) [inline src/main/webapp/i18n/de/userSettings.json:60]:
```suggestion
                "fingerprintsExplanation": "Mit SSH-Schlüsseln kannst du eine sichere Verbindung zwischen deinem Computer und Artemis herstellen. SSH-Fingerabdrücke stellen sicher, dass du eine Verbindung zum richtigen Gerät herstellt.",
```

File: src/main/webapp/i18n/en/userSettings.json [hunk 3]
```diff
@@ -55,6 +55,10 @@
                 "expiresOn": "Expires on",
                 "hasExpiredOn": "Expired on",
                 "fingerprint": "Fingerprint",
+                "fingerprints": "Show SSH fingerprints",
+                "sshFingerprints": "SSH Fingerprints",
+                "fingerprintsExplanation": "SSH keys allow you to establish a secure connection between your computer and Artemis. SSH fingerprints verify that the client is connecting to the correct host.",
```
Comment (iyannsch) [inline src/main/webapp/i18n/en/userSettings.json:60]:
```suggestion
                "fingerprintsExplanation": "SSH keys allow you to establish a secure connection between your computer and Artemis. SSH fingerprints verify that you connect to the correct device.",
```

**Review by iyannsch — CHANGES_REQUESTED** [review]
Looking good :) I added two small CRs regarding the i18n

File: src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/ssh/SshFingerprintsProviderService.java (hunk 1 above)
Comment (SimonEntholzer) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/ssh/SshFingerprintsProviderService.java:46]: loadKeys returns the list of keys of the SSH server, the argument is the session. We can set it to null here as we do not actually have an active session context in this place. We also don't need it here

**Review by raffifasaro — APPROVED** [review]
code LGTM

**Review by cremertim — APPROVED** [review]
Tested according to the steps, works as expected

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Works as described

image

**Review by az108 — APPROVED** [review]
Code LGTM 👍

File: src/main/java/de/tum/cit/aet/artemis/programming/web/localvc/ssh/SshFingerprintsProviderResource.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    /**
+     * GET /ssh-fingerprints
+     *
+     * @return the SSH fingerprints for the keys a user uses
+     */
+    @GetMapping(value = "ssh-fingerprints", produces = MediaType.APPLICATION_JSON_VALUE)
+    @EnforceAtLeastStudent
+    @FeatureToggle(Feature.Exports)
```
Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/localvc/ssh/SshFingerprintsProviderResource.java:40]: I don't think this feature toggle makes sense here

**Review by magkue — DISMISSED** [review]

**Review by iyannsch — DISMISSED** [review]

**Review by coolchock — APPROVED** [review]
re-approve after FeatureToggle was removed

**Review by BBesrour — APPROVED** [review]

**Review by muradium — APPROVED** [review]
Tested on TS6, fingerprint is displayed correctly

**Review by magkue — APPROVED** [review]
Reapprove

**Review by HawKhiem — APPROVED** [review]
Tested on TS3. Reapprove

image

**Review by krusche — APPROVED** [review]

**Pull request merged**